pages/html/components.md

```diff
@@ -0,0 +1,119 @@
+---
+title: Components
+---
+
+# Components
+
+You can build reusable & composable Phlex views.
+
+For example, you may need to define multiple sections (slots) in a view. This can be accomplished by defining public instance methods on the view that accept blocks:
+
+```phlex
+example do |e|
+  e.tab "card.rb", <<~RUBY
+    class Card < Phlex::HTML
+      def template(&)
+        article(class: "card", &)
+      end
+
+      def title(&)
+        div(class: "title", &)
+      end
+
+      def body(&)
+        div(class: "body", &)
+      end
+    end
+  RUBY
+
+  e.tab "example.rb", <<~RUBY
+    class Example < Phlex::HTML
+      def template
+        render Card.new do |card|
+          card.title do
+            h1 { "Title" }
+          end
+
+          card.body do
+            p { "Body" }
+          end
+        end
+      end
+    end
+  RUBY
+
+  e.execute "Example.new.call"
+end
+```
+
+This would work just fine for a list of views as each method can be called multiple times.
+
+One caveat of defining the view this way is `title` and `body` could be called in any order. This offers flexibility, but what if you need to make sure your markup is output in a consistent order?
+
+For this level of control you can `yield` early from your `template`, then use the public methods to save the blocks, passing them to back to your template at render time.
+
+```phlex
+example do |e|
+  e.tab "list.rb", <<~RUBY
+    class List < Phlex::HTML
+      def initialize
+        @items = []
+      end
+
+      def template(&)
+        yield(self)
```

Comment:
There’s an explicit pattern for this that hasn't been documented yet, but essentially you just include `DeferredRender` and then your `template` doesn't receive a block as it’s yielded and thrown away early.

https://github.com/joeldrapper/phlex/blob/15a3f907c937df2f29c94167f4e28cad0da4efe3/lib/phlex/html.rb#L78-L80

I think this is better than yielding manually because otherwise you can end up content appearing above your component. Deferred rendering vanishes the block by temporarily setting the target buffer to `BlackHole`, which is a special object that fits the buffer interface but does nothing.

https://github.com/joeldrapper/phlex/blob/15a3f907c937df2f29c94167f4e28cad0da4efe3/lib/phlex/html.rb#L198-L199